File: app/src/main/java/org/alkaline/taskbrain/dsl/runtime/Executor.kt
package org.alkaline.taskbrain.dsl.runtime

import org.alkaline.taskbrain.dsl.language.Assignment
import org.alkaline.taskbrain.dsl.language.CallExpr
import org.alkaline.taskbrain.dsl.language.CurrentNoteRef
import org.alkaline.taskbrain.dsl.language.Directive
import org.alkaline.taskbrain.dsl.language.Expression
import org.alkaline.taskbrain.dsl.language.MethodCall
import org.alkaline.taskbrain.dsl.language.NumberLiteral
import org.alkaline.taskbrain.dsl.language.PatternExpr
import org.alkaline.taskbrain.dsl.language.PropertyAccess
import org.alkaline.taskbrain.dsl.language.StatementList
import org.alkaline.taskbrain.dsl.language.StringLiteral
import org.alkaline.taskbrain.dsl.language.VariableRef

/**
 * Evaluates DSL expressions and produces runtime values.
 *
 * Milestone 3: Supports parenthesized calls with named arguments.
 * Milestone 4: Supports pattern expressions.
 * Milestone 6: Supports current note reference and property access.
 * Milestone 7: Supports assignment, statement lists, variables, and method calls.
 */
class Executor {

    /**
     * Execute a directive and return its result.
     */
    fun execute(directive: Directive, env: Environment = Environment()): DslValue {
        return evaluate(directive.expression, env)
    }

    /**
     * Evaluate an expression to produce a value.
     *
     * Milestone 7: Added Assignment, StatementList, VariableRef, MethodCall.
     */
    fun evaluate(expr: Expression, env: Environment): DslValue {
        return when (expr) {
            is NumberLiteral -> NumberVal(expr.value)
            is StringLiteral -> StringVal(expr.value)
            is CallExpr -> evaluateCall(expr, env)
            is PatternExpr -> evaluatePattern(expr)
            is CurrentNoteRef -> evaluateCurrentNoteRef(expr, env)
            is PropertyAccess -> evaluatePropertyAccess(expr, env)
            is Assignment -> evaluateAssignment(expr, env)
            is StatementList -> evaluateStatementList(expr, env)
            is VariableRef -> evaluateVariableRef(expr, env)
            is MethodCall -> evaluateMethodCall(expr, env)
        }
    }

    /**
     * Evaluate a current note reference.
     * Returns the current note from the environment.
     *
     * Milestone 6.
     */
    private fun evaluateCurrentNoteRef(expr: CurrentNoteRef, env: Environment): NoteVal {
        return env.getCurrentNote()
            ?: throw ExecutionException(
                "No current note in context (use [.] only within a note)",
                expr.position
            )
    }

    /**
     * Evaluate property access on an expression.
     * Example: note.path, note.created
     *
     * Milestone 6.
     */
    private fun evaluatePropertyAccess(expr: PropertyAccess, env: Environment): DslValue {
        val target = evaluate(expr.target, env)
        return when (target) {
            is NoteVal -> target.getProperty(expr.property, env)
            else -> throw ExecutionException(
                "Cannot access property '${expr.property}' on ${target.typeName}",
                expr.position
            )
        }
    }

    /**
     * Evaluate an assignment expression.
     * Example: [x: 5], [.path: "foo"]
     *
     * Milestone 7.
     */
    private fun evaluateAssignment(expr: Assignment, env: Environment): DslValue {
        val value = evaluate(expr.value, env)

        when (val target = expr.target) {
            is VariableRef -> {
                // Variable definition: [x: 5]
                env.define(target.name, value)
            }
            is PropertyAccess -> {
                // Property assignment: [.path: "foo"] or [note.path: "bar"]
                val targetObj = evaluate(target.target, env)
                when (targetObj) {
                    is NoteVal -> {
                        targetObj.setProperty(target.property, value, env)
                    }
                    else -> throw ExecutionException(
                        "Cannot assign to property '${target.property}' on ${targetObj.typeName}",
                        expr.position
                    )
                }
            }
            is CurrentNoteRef -> {
                // [.: value] - not really meaningful, but could be used to replace note
                throw ExecutionException(
                    "Cannot assign directly to current note. Use property assignment like [.path: value]",
                    expr.position
                )
            }
            else -> throw ExecutionException(
                "Invalid assignment target",
                expr.position
            )
        }

        return value
    }

    /**
     * Evaluate a statement list, returning the value of the last statement.
     * Example: [x: 5; y: 10; add(x, y)]
     *
     * Milestone 7.
     */
    private fun evaluateStatementList(expr: StatementList, env: Environment): DslValue {
        var result: DslValue = StringVal("")  // Default if empty
        for (statement in expr.statements) {
            result = evaluate(statement, env)
        }
        return result
    }

    /**
     * Evaluate a variable reference.
     * Example: [x] where x was previously defined
     *
     * Milestone 7.
     */
    private fun evaluateVariableRef(expr: VariableRef, env: Environment): DslValue {
        return env.get(expr.name)
            ?: throw ExecutionException(
                "Undefined variable '${expr.name}'",
                expr.position
            )
    }

    /**
     * Evaluate a method call on an expression.
     * Example: [.append("text")], [note.append("text")]
     *
     * Milestone 7.
     */
    private fun evaluateMethodCall(expr: MethodCall, env: Environment): DslValue {
        val target = evaluate(expr.target, env)

        // Evaluate arguments
        val positionalArgs = expr.args.map { evaluate(it, env) }
        val namedArgs = expr.namedArgs.associate { it.name to evaluate(it.value, env) }
        val args = Arguments(positionalArgs, namedArgs)

        // Dispatch to the appropriate method based on target type
        return when (target) {
            is NoteVal -> target.callMethod(expr.methodName, args, env, expr.position)
            else -> throw ExecutionException(
                "Cannot call method '${expr.methodName}' on ${target.typeName}",
                expr.position
            )
        }
    }

    /**
     * Evaluate a pattern expression to produce a PatternVal.
     */
    private fun evaluatePattern(expr: PatternExpr): PatternVal {
        return PatternVal.compile(expr.elements)
    }

    /**
     * Evaluate a function call expression.
     *
     * Milestone 7: For zero-arg calls, first check if it's a variable reference.
     */
    private fun evaluateCall(expr: CallExpr, env: Environment): DslValue {
        // For zero-arg calls, first check if this is a variable reference
        // This allows [x: 5; x] where second 'x' refers to the variable
        if (expr.args.isEmpty() && expr.namedArgs.isEmpty()) {
            val variableValue = env.get(expr.name)
            if (variableValue != null) {
                return variableValue
            }
        }

        val function = BuiltinRegistry.get(expr.name)
            ?: throw ExecutionException(
                "Unknown function or variable '${expr.name}'",
                expr.position
            )

        // Evaluate positional arguments
        val positionalValues = expr.args.map { evaluate(it, env) }

        // Evaluate named arguments
        val namedValues = expr.namedArgs.associate { namedArg ->
            namedArg.name to evaluate(namedArg.value, env)
        }

        // Build Arguments container
        val args = Arguments(positionalValues, namedValues)

        // Call the function
        return try {
            function.call(args, env)
        } catch (e: ExecutionException) {
            throw e
        } catch (e: Exception) {
            throw ExecutionException(
                "Error in '${expr.name}': ${e.message}",
                expr.position
            )
        }
    }
}

/**
 * Exception thrown during DSL execution.
 */
class ExecutionException(
    message: String,
    val position: Int? = null
) : RuntimeException(message)

File: web/src/dsl/runtime/Executor.ts
import type { Directive, Expression } from '../language/Expression'
import { containsDynamicCalls } from '../language/DynamicCallAnalyzer'
import type { DslValue, LambdaVal } from './DslValue'
import {
  numberVal,
  stringVal,
  lambdaVal,
  compilePattern,
  typeName,
} from './DslValue'
import { Arguments } from './Arguments'
import { Environment } from './Environment'
import { ExecutionException } from './ExecutionException'
import { BuiltinRegistry } from './BuiltinRegistry'
import { callMethod } from './MethodHandler'
import * as NotePropertyHandler from './NotePropertyHandler'
import { callNoteMethod } from './NoteMethodHandler'

/**
 * Evaluates DSL expressions and produces runtime values.
 */
export class Executor {
  /**
   * Execute a directive and return its result.
   */
  execute(directive: Directive, env: Environment = Environment.create()): DslValue {
    const execEnv = this.createExecutionEnvironment(env)
    const result = this.evaluate(directive.expression, execEnv)

    // Validate that temporal values are wrapped in once[...] or refresh[...]
    this.validateTemporalResult(result, directive.expression, directive.startPosition)

    return result
  }

  /**
   * Evaluate an expression to produce a value.
   */
  evaluate(expr: Expression, env: Environment): DslValue {
    switch (expr.kind) {
      case 'NumberLiteral': return numberVal(expr.value)
      case 'StringLiteral': return stringVal(expr.value)
      case 'CallExpr': return this.evaluateCall(expr, env)
      case 'PatternExpr': return compilePattern(expr.elements)
      case 'CurrentNoteRef': return this.evaluateCurrentNoteRef(expr, env)
      case 'PropertyAccess': return this.evaluatePropertyAccess(expr, env)
      case 'Assignment': return this.evaluateAssignment(expr, env)
      case 'StatementList': return this.evaluateStatementList(expr, env)
      case 'VariableRef': return this.evaluateVariableRef(expr, env)
      case 'MethodCall': return this.evaluateMethodCall(expr, env)
      case 'LambdaExpr': return this.evaluateLambda(expr, env)
      case 'LambdaInvocation': return this.evaluateLambdaInvocation(expr, env)
      case 'OnceExpr': return this.evaluateOnce(expr, env)
      case 'RefreshExpr': return this.evaluateRefresh(expr, env)
    }
  }

  /**
   * Invoke a lambda with the given arguments.
   */
  invokeLambda(lambda: LambdaVal, args: DslValue[]): DslValue {
    const localEnv = (lambda.capturedEnv as Environment).child()
    lambda.params.forEach((param, i) => {
      localEnv.define(param, args[i]!)
    })
    return this.evaluate(lambda.body, localEnv)
  }

  // ---- Private evaluation methods ----

  private createExecutionEnvironment(baseEnv: Environment): Environment {
    if (baseEnv.getExecutor()) return baseEnv
    return baseEnv.withExecutor(this)
  }

  private validateTemporalResult(result: DslValue, expr: Expression, position: number): void {
    if (!containsDynamicCalls(expr)) return
    if (result.kind === 'DateVal' || result.kind === 'TimeVal' || result.kind === 'DateTimeVal') {
      if (expr.kind !== 'OnceExpr' && expr.kind !== 'RefreshExpr') {
        throw new ExecutionException(
          `Bare ${typeName(result)} value is not allowed. Use once[${typeName(result)}] to capture a snapshot, or refresh[...] to specify when to update.`,
          position,
        )
      }
    }
  }

  private evaluateLambda(expr: Expression & { kind: 'LambdaExpr' }, env: Environment): DslValue {
    return lambdaVal(expr.params, expr.body, env.capture())
  }

  private evaluateLambdaInvocation(expr: Expression & { kind: 'LambdaInvocation' }, env: Environment): DslValue {
    const lv = this.evaluateLambda(expr.lambda, env) as LambdaVal
    const argValues = expr.args.map((a) => this.evaluate(a, env))
    if (argValues.length !== lv.params.length) {
      throw new ExecutionException(`Lambda requires ${lv.params.length} argument(s), got ${argValues.length}`, expr.position)
    }
    if (expr.namedArgs.length > 0) {
      throw new ExecutionException('Named arguments not supported for lambda invocation', expr.position)
    }
    return this.invokeLambda(lv, argValues)
  }

  private evaluateOnce(expr: Expression & { kind: 'OnceExpr' }, env: Environment): DslValue {
    const cacheKey = `once:${hashCode(expr.body)}`
    const cache = env.getOrCreateOnceCache()
    const cached = cache.get(cacheKey)
    if (cached) return cached
    const result = this.evaluate(expr.body, env)
    cache.put(cacheKey, result)
    return result
  }

  private evaluateRefresh(expr: Expression & { kind: 'RefreshExpr' }, env: Environment): DslValue {
    return this.evaluate(expr.body, env)
  }

  private evaluateCurrentNoteRef(expr: Expression & { kind: 'CurrentNoteRef' }, env: Environment): DslValue {
    const note = env.getCurrentNote()
    if (!note) throw new ExecutionException('No current note in context (use [.] only within a note)', expr.position)
    return note
  }

  private evaluatePropertyAccess(expr: Expression & { kind: 'PropertyAccess' }, env: Environment): DslValue {
    const target = this.evaluate(expr.target, env)
    if (target.kind === 'NoteVal') return NotePropertyHandler.getProperty(target, expr.property, env)
    throw new ExecutionException(`Cannot access property '${expr.property}' on ${typeName(target)}`, expr.position)
  }

  private evaluateAssignment(expr: Expression & { kind: 'Assignment' }, env: Environment): DslValue {
    const value = this.evaluate(expr.value, env)

    if (expr.target.kind === 'VariableRef') {
      env.define(expr.target.name, value)
    } else if (expr.target.kind === 'PropertyAccess') {
      const targetObj = this.evaluate(expr.target.target, env)
      if (targetObj.kind === 'NoteVal') {
        // Note: In the web port, setProperty is async but we call it synchronously here
        // for compatibility with the synchronous evaluation model.
        // The async operations will be handled at a higher level.
        NotePropertyHandler.setProperty(targetObj, expr.target.property, value, env)
      } else {
        throw new ExecutionException(`Cannot assign to property '${expr.target.property}' on ${typeName(targetObj)}`, expr.position)
      }
    } else if (expr.target.kind === 'CurrentNoteRef') {
      throw new ExecutionException('Cannot assign directly to current note. Use property assignment like [.path: value]', expr.position)
    } else {
      throw new ExecutionException('Invalid assignment target', expr.position)
    }

    return value
  }

  private evaluateStatementList(expr: Expression & { kind: 'StatementList' }, env: Environment): DslValue {
    let result: DslValue = stringVal('')
    for (const statement of expr.statements) {
      result = this.evaluate(statement, env)
    }
    return result
  }

  private evaluateVariableRef(expr: Expression & { kind: 'VariableRef' }, env: Environment): DslValue {
    const val = env.get(expr.name)
    if (val === null) throw new ExecutionException(`Undefined variable '${expr.name}'`, expr.position)
    return val
  }

  private evaluateMethodCall(expr: Expression & { kind: 'MethodCall' }, env: Environment): DslValue {
    const target = this.evaluate(expr.target, env)
    const positionalArgs = expr.args.map((a) => this.evaluate(a, env))
    const namedArgs = new Map(expr.namedArgs.map((na) => [na.name, this.evaluate(na.value, env)]))
    const args = new Arguments(positionalArgs, namedArgs)

    if (target.kind === 'NoteVal') {
      // Note: callNoteMethod is async for append/setProperty, but we call it here.
      // The async result is a Promise<DslValue> which would need to be awaited at a higher level.
      return callNoteMethod(target, expr.methodName, args, env, expr.position) as unknown as DslValue
    }
    return callMethod(target, expr.methodName, args, expr.position)
  }

  private evaluateCall(expr: Expression & { kind: 'CallExpr' }, env: Environment): DslValue {
    // Check if this is a variable
    const variableValue = env.get(expr.name)
    if (variableValue !== null) {
      if (variableValue.kind === 'LambdaVal') {
        const argValues = expr.args.map((a) => this.evaluate(a, env))
        if (argValues.length !== variableValue.params.length) {
          throw new ExecutionException(`Lambda requires ${variableValue.params.length} argument(s), got ${argValues.length}`, expr.position)
        }
        if (expr.namedArgs.length > 0) {
          throw new ExecutionException('Named arguments not supported for lambda invocation', expr.position)
        }
        return this.invokeLambda(variableValue, argValues)
      }
      // Non-lambda variable
      if (expr.args.length === 0 && expr.namedArgs.length === 0) return variableValue
      throw new ExecutionException(`Cannot call ${typeName(variableValue)} as a function`, expr.position)
    }

    // Look up builtin function
    const fn = BuiltinRegistry.get(expr.name)
    if (!fn) throw new ExecutionException(`Unknown function or variable '${expr.name}'`, expr.position)

    const positionalValues = expr.args.map((a) => this.evaluate(a, env))
    const namedValues = new Map(expr.namedArgs.map((na) => [na.name, this.evaluate(na.value, env)]))
    const args = new Arguments(positionalValues, namedValues)

    try {
      return fn.call(args, env) as DslValue
    } catch (e) {
      if (e instanceof ExecutionException) throw e
      throw new ExecutionException(`Error in '${expr.name}': ${(e as Error).message}`, expr.position)
    }
  }
}

/**
 * Simple hash code for an expression (used for once[] cache keys).
 */
function hashCode(expr: Expression): number {
  const str = JSON.stringify(expr)
  let hash = 0
  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i)
    hash = ((hash << 5) - hash + char) | 0
  }
  return hash
}
